### DISCUSSION TITLE
Npm 7 and Storybook monorepo

### DISCUSSION THREAD

**Discussion by dwightjack** [renovatebot/renovate#12442] [Request Help]
Hi,

We are using renovate on a project using Storybook and npm@7. Renovate PRs are failing due to a `Conflicting peer dependency` error. 

Setting `--legacy-peer-deps` in `.npmrc` works, but I am wondering if there is a better way to solve this issue.

### Package.json config

```json
"devDependencies": {
    "@storybook/addon-a11y": "6.3.8",
    "@storybook/addon-essentials": "6.3.8",
    "@storybook/addon-postcss": "2.0.0",
    "@storybook/addon-storysource": "6.3.8",
    "@storybook/vue": "6.3.8"
}
```

### Renovate config

```json5
{
  extends: [
    'config:js-app',
    ':automergePatch',
    ':label(Renovate)',
    ':prConcurrentLimit10',
    ':prNotPending',
  ],
  packageRules: [
    {
      addLabels: ['dependencies'],
      matchDepTypes: ['dependencies'],
    },
    {
      addLabels: ['devDependencies'],
      matchDepTypes: ['devDependencies'],
    }
  ],
}
```

### Demo PR

![image](https://user-images.githubusercontent.com/104721/139796413-3dbd25f6-918b-4a7c-8ea4-cff73d569a95.png)

### Error

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: <project_name>
npm ERR! Found: @storybook/addon-essentials@6.3.8
npm ERR! node_modules/@storybook/addon-essentials
npm ERR!   dev @storybook/addon-essentials@"6.3.12" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! dev @storybook/addon-essentials@"6.3.12" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: @storybook/vue@6.3.12
npm ERR! node_modules/@storybook/vue
npm ERR!   peerOptional @storybook/vue@"6.3.12" from @storybook/addon-essentials@6.3.12
npm ERR!   node_modules/@storybook/addon-essentials
npm ERR!     dev @storybook/addon-essentials@"6.3.12" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /tmp/renovate-cache/others/npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /tmp/renovate-cache/others/npm/_logs/2021-11-02T03_39_33_026Z-debug.log
```

**Comment by viceice** [comment]
There is nothing renovate can do here 🤷‍♂️

**Comment by rarkins** [comment]
Sounds like there's a `@storybook/addon-essentials@6.3.8` somewhere in the tree?

First try to reproduce the problem locally. Check out Renovate's branch and then run `npm install` (ideally delete `node_modules` first).

Assuming that fils, then remove both the `package-lock.json` and `node_modules` completely before running `npm install`, does it succeed?

**Reply by dwightjack** [reply]
@rarkins thank you for your reply. 

Removing `package-lock.json` worked locally. The PR from Renovate changed only `package.json` but didn't update the lock file (I guess because of that  `peerDependencies`error ?), so creating a fresh one solved the issue. 

`@storybook/addon-essentials` is in the `devDependencies` of our project. Looking at the storybook repo, the package has `@storybook/vue` as optional peer dep (https://github.com/storybookjs/storybook/blob/v6.3.12/addons/essentials/package.json#L63-L76).

I don't know how Renovate works internally, but it looks like that it is trying to install the updated packages one by one, so that when it runs `npm i @storybook/addon-essentials@6.13.12` it encounters a wrong peer dep `@storybook/vue@6.8.3`

I guess that updating  `@storybook/vue` first and then `@storybook/addon-essentials` might solve the issue. Is there any way to enforce a specific update order?

**Reply by rarkins** [reply]
What's the exact `docker run` command that's run that fails?

**Reply by dwightjack** [reply]
Sorry, I should have mentioned that we are using the renovatebot on Github. Where should I look for the docker command?

**Reply by viceice** [reply]
https://app.renovatebot.com/dashboard

link is in pr body too

**Reply by dwightjack** [reply]
In the latest job log I can see a `DEBUG: rawExec err` reporting the issue I encountered. The related command is:

```
docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/repo-name\":\"/mnt/renovate/gh/repo-name\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/mnt/renovate/gh/repo-name\" docker.io/renovate/node:v16.9.1 bash -l -c \"npm i -g 'npm@>=8.1.0' || true && hash -d npm && npm install --package-lock-only --no-audit --ignore-scripts\"
```

**Reply by rarkins** [reply]
Why did you say, "it looks like that it is trying to install the updated packages one by one"?

**Reply by dwightjack** [reply]
Sorry, it was just a feeling. Not knowing how Renovate actually works I had the feeling that the steps it was doing were:

- check for updates
- verify that `@storybook/*` packages can be updated
- start running npm install for each package `npm install @storybook/addons-essentials`
- error

But, looking at the log I guess it first updates `package.json` and then runs `npm install --package-lock-only --no-audit --ignore-scripts`. Am I correct?

**Reply by rarkins** [reply]
Yes, it manually applies updates to `package.json` and then calls `npm install` just once.

We want to find a way we can coerce npm into getting it to work, but wiping the package lock completely is undesirable because it might result in a lot of unexpected and unrelated changes.

Can you reproduce this problem in a public repo we can test against?

**Reply by dwightjack** [reply]
Yes, I can try to reproduce the issue. I will be away tomorrow, so I might be getting back to you in a couple of days 😃.

**Reply by dwightjack** [reply]
@rarkins I created a demo repository to reproduce the issue: https://github.com/dwightjack/vue-cli-storybook-demo.

Here is the Renovate PR related to Storybook: https://github.com/dwightjack/vue-cli-storybook-demo/pull/2

Let me know if you need anything.

**Reply by dwightjack** [reply]
Also, I can grant you write permission to the repo if you need it.

**Reply by rarkins** [reply]
I've forked it, thanks.

My notes:
- Checked out `renovate/patch-storybook-monorepo`
- Can replicate problem using `npm install --package-lock-only --no-audit --ignore-scripts`
- Also verified that it replicates using `npm install` alone
- Running `npm i --save-dev @storybook/addon-essentials@6.3.12 @storybook/vue@6.3.12` also fails, including if you roll back the manual changes to `package.json` first
- Removing `package-lock.json` and then running `npm install --package-lock-only --no-audit --ignore-scripts` works. Gist of diff: https://gist.github.com/rarkins/8738cc73cfb906606c778a839a93ff7d
- If I manually delete all `@storybook/` entries from `package-lock.json` - and there are hundreds - and then run `npm install --package-lock-only --no-audit --ignore-scripts` then it works. Diff: https://gist.github.com/rarkins/bdf2c88b479fe72929c9adc1e62d6088. The problem with that is that a "machine" (Renovate) can't necessarily know which types of packages to delete unless it was to construct a tree and work its way down from there.
- If I instead just delete the "node_modules/x" and "node_modules/x/*" entries - where x is an upgraded dependency - then it seems to work!

Maybe for now the most simplest approach in such a scenario would be to catch the initial failure, remove the lock file, and if it works with no lock file present then commit that - even though ideally we wouldn't shake the whole lock file. The best approach - although not necessarily guaranteed to work - would be to remove the node_modules entries.

**Reply by rarkins** [reply]
Fix in progress:  #12485

**Reply by dwightjack** [reply]
Thanks! 🙇
